flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/DefaultLeaderRetrievalService.java

```diff
@@ -0,0 +1,155 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.leaderretrieval;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.function.SupplierWithException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * The counterpart to the {@link org.apache.flink.runtime.leaderelection.DefaultLeaderElectionService}.
+ * Composed with different {@link LeaderRetrievalDriver}, we could retrieve the leader information from
+ * different storage. The leader address as well as the current leader session ID will be retrieved from
+ * {@link LeaderRetrievalDriver}.
+ */
+public class DefaultLeaderRetrievalService implements LeaderRetrievalService {
+	private static final Logger LOG = LoggerFactory.getLogger(DefaultLeaderRetrievalService.class);
+
+	private final Object lock = new Object();
+
+	private String lastLeaderAddress;
+
+	private UUID lastLeaderSessionID;
+
+	private volatile boolean running;
+
+	/** Listener which will be notified about leader changes. */
+	private volatile LeaderRetrievalListener leaderListener;
+
+	private final LeaderRetrievalDriver leaderRetrievalDriver;
+
+	private final StateHandler leaderRetrievalStateHandler;
+
+	/**
+	 * Creates a default leader retrieval service with specified {@link LeaderRetrievalDriver}.
+	 *
+	 * @param leaderRetrievalDriver {@link LeaderRetrievalDriver} implemented by Zookeeper, Kubernetes, etc.
+	 */
+	public DefaultLeaderRetrievalService(LeaderRetrievalDriver leaderRetrievalDriver) {
+		this.leaderRetrievalDriver = leaderRetrievalDriver;
+
+		this.lastLeaderAddress = null;
+		this.lastLeaderSessionID = null;
+
+		running = false;
+
+		this.leaderRetrievalStateHandler = new StateHandler();
+	}
+
+	@Override
+	public void start(LeaderRetrievalListener listener) throws Exception {
+		Preconditions.checkNotNull(listener, "Listener must not be null.");
+		Preconditions.checkState(leaderListener == null, "DefaultLeaderRetrievalService can " +
+			"only be started once.");
+
+		LOG.info("Starting DefaultLeaderRetrievalService with {}.", leaderRetrievalDriver);
+
+		synchronized (lock) {
+			leaderListener = listener;
+			leaderRetrievalDriver.start(leaderRetrievalStateHandler);
+
+			running = true;
+		}
+	}
+
+	@Override
+	public void stop() throws Exception {
+		synchronized (lock) {
+			if (!running) {
+				return;
+			}
+			running = false;
+
+			LOG.info("Stopping DefaultLeaderRetrievalService {} with {}.", leaderRetrievalDriver);
+
+			leaderRetrievalDriver.stop();
+		}
+	}
+
+	/**
+	 * Helper class for the specific {@link LeaderRetrievalDriver} to notify the leader changes.
+	 */
+	public class StateHandler {
```

Comment:
I would suggest to call it `LeaderRetrievalEventHandler` and make it an interface for easier testability.